core/src/main/java/org/apache/iceberg/TableOperations.java

```diff
@@ -117,7 +117,7 @@ default long newSnapshotId() {
     UUID uuid = UUID.randomUUID();
     long mostSignificantBits = uuid.getMostSignificantBits();
     long leastSignificantBits = uuid.getLeastSignificantBits();
-    return Math.abs(mostSignificantBits ^ leastSignificantBits);
+    return (mostSignificantBits ^ leastSignificantBits) & Long.MAX_VALUE;
```

Comment:
I'm curious whether are there two different `UUID` whose `Math.abs(mostSignificantBits ^ leastSignificantBits)`  are the same.  If so, then we will encounter a duplicated new snapshot id which could be a serious bug ?

Reply:
There definitely are :), 

There are at least 2 (same is with the ^ Long.MaxValue)
We have 64 random bits ^ 64 Random bits and we put it into 63 bits (Remember the first bit is sign so we only have 63 bits of values after abs) BUT because of the overflow we get 1 extra value in our output domain, so we actually get 63 Bits (All positive longs) + 1 (Long.minValue)

With  the 
^ & method
We actually do the same thing
64 random bits & 64 random bits ^ 63 bits (the first bit is 0 because Long.MaxValue is positive) so we still can only get 63 bit's worth of values in there. (No Extra 1 for long.min value)

I think the only difference is that because of the 0 point you end up with 3 values which end up matching Long.maxValue: Long.maxValue, Long.minValue, -Long.maxValue

Now again this basically isn't an issue any more than using a UUID is for repeated values since the chance of collision is really small so no big worry.

Reply:
TLDR; we have collisions at a rate of 1 /  (Long.Domain / 2)

Reply:
Okay, after read the original PR https://github.com/apache/iceberg/pull/57,  although there's small probability to produce the same snapshot id ,  it won't be able to commit the iceberg table metadata because we will encounter a non-recoverable IllegalArgumentException when building the [immutable `snapshotsById` map ](https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/core/src/main/java/org/apache/iceberg/TableMetadata.java#L888).  So this should not be a bug.

( I also did a small test to validate the conflicts in `ImmutableMap`:

```java
public static void main(String[] args) {
    ImmutableMap.Builder<Long, String> builder = ImmutableMap.builder();
    builder.put(1L, "hello");
    builder.put(1L, "world");
    Map<Long, String> map = builder.build();
    System.out.println(map);
  }
``` 

It will throw a `IllegalArgumentException`:

```
Exception in thread "main" java.lang.IllegalArgumentException: Multiple entries with same key: 1=world and 1=hello
	at org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:214)
	at org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:208)
	at org.apache.iceberg.relocated.com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:146)
	at org.apache.iceberg.relocated.com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:109)
	at org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:392)
	at org.apache.iceberg.TestImmutableMap.main(TestImmutableMap.java:33)
```


)